packages/jsii-diff/README.md

```diff
@@ -0,0 +1,82 @@
+# jsii-diff
+
+__jsii-diff__ compares two jsii assemblies for compatibility.
+
+In the future, it will be able to do generic comparisons, but for
+now it will compare assemblies for API compatibility, and exit
+with a non-zero exit code if any **stable** APIs have had incompatible
+changes.
+
+API items that have no stability are treated as **experimental** (i.e.,
+do not affect API compatibility.
+
+## Usage
+
+To compare two JSII packages:
+
+    jsii-diff <path/to/old> [path/to/new]
+
+To compare current package against latest published NPM release:
+
+    # Can leave out . for current directory
+    jsii-diff npm:package [.]
+
+## Details
+
+__jsii-diff__ will assert that code written against version **A** of a library
+will still typecheck when compiled against version **B** of that library. It
+does this by verifying the following properties:
+
+- Any type (class/interface/enum) in **A** must also exist in **B**.
+- Enums have only added members.
+- Classes and behavioral interfaces have only added members, restricted
+  property types and method return types, and only widened argument types.
+
+### Subclassing
+
+In principle, in order to stay API compatible no members can ever be added to
+interfaces, and no abstract members can ever be added to classes. This is
+because any user of the library could be implementing the interface or could
+be inheriting frmo the class. In those cases, the addition of an abstract
+member would break their code, since they wouldn't be implementing that
+member yet.
+
+Because this contract is a little too restrictive for evolution of libraries,
+we assume most interfaces will only be implemented by library classes, so we
+can disregard burden on implementors as part of API compatibility evaluation.
+
+If an interface or class is intended to be subclassed by library users,
```

Comment:
User implementable interfaces are quite rare in our library definition, so I would rather annotate the exceptions than the rule.

Also, while C# can `seal` classes, sealing interfaces does not make any sense.

Reply:
The problem is that having a relaxed constraint for interface breakage without any signal to users is a bad API. Users *can* implement any of our interfaces and they will be broken without warning. If, however, we require that, for example, all interfaces will have a $version field or something like that people will understand that this may change...

Reply:
Not sure what you're proposing. Annotate all of our classes and interfaces with a `@do_not_extend` annotation? (not literally that of course, but you get my drift)

That has the problem that it can be forgotten, so I guess now all classes and  interfaces will have to be annotated with `@open_for_subclassing true|false`, which is ever more annotation burden on library authors.

`@version`ing in the docstring brings exactly the same burden,  and versioning outside the docstring (in the code) leads to:

```ts
interface IConstruct2 extends IConstruct {
  // ...
}
```

Which can be done anyway but is not a great experience.

Is saying "none of this API surface can be subclassed without stability guarantee unless we tell you to" not a pragmatic compromise? It's not like we're physically preventing people from doing it, they should just be aware they may need to modify their code when the upgrade to a new version.